README.md
# nhlscrape
A package for scraping the nhl API and HTML reports to build a database. Can be used to generate statistics from the information pulled.

## Usage
To use this package:
1. Import 'nhlscraper' and run SetDbPath to set the location for the database file.
2. Use AddAllTeams() and the find team_id using the GetTeamId function.
3. Find game_ids using the GetGameIdRange() function.
4. Add game reports to the database using AddGameEvents().
5. Find a player_id using the GetPlayerId('player_name') function.
6. Get player statistics using GetPlayerStats(). You can also manually query the database using QueryDb.


Example:
```
library(nhlscrape)

# Set db path to somewhere
SetDbPath("C:/Users/Adam/Documents/nhl.sqlite")

# Select the leafs
AddAllTeamsDb()
team_id <- GetTeamId("toronto maple leafs")
gids <- GetGameIdRange(team_id, "2019-09-30", "2019-12-18")

# Add games
AddGameEvents(gids)

# Get stats for player
# Tavares
player_id <- GetPlayerId("john tavares")

stats <- GetPlayerStats(player_id, gids, team_id)
```

## Example Shiny App
You can see an example of the heatmap function being plotted in this demo shiny app at:
https://aazoulay.shinyapps.io/nhl_shiny/


NEWS.md
# Version 0.1.3

* Added support for emergency players subbing into the game
* Fixed outliers on heatmap by using average shot position to find offensive zone

# Version 0.1.2

* Updated HTML links in the description
* Updated date in license file
* Changed the examples to use an example database file in a temp directory
* Changed \dontrun to \donttest on api calls in the examples

# Version 0.1.1

* Shortened title
* Updated title and description with proper api notation
* Added https reference links in description
* Changed examples to not do any local writing
* Changed from options to private environment for global variables
* Added an error preventing the user from writing to database unless they manually run the SetDbPath command
* Added new functions for getting transformed heatmap coordinates and more advanced stats

# Version 0.1.0

* Initial submission
* Add game events to database from 'NHL' api
* Generate reports for game statistics


tests/testthat.R
library(testthat)
library(nhlscrape)

test_check("nhlscrape")


tests/teststats.R
context("Statistics equal")
library(nhlscrape)

test_that("Statistics for even strength and all situtations is the same", {
  expect_equal(GetPlayerStats(player_id, game_id, team_id)$CF[1], 28)
  expect_equal(GetPlayerStats(player_id, game_id, team_id)$CA[1], 12)
  expect_equal(GetPlayerStats(player_id, game_id, team_id)$CF[2], 13)
  expect_equal(GetPlayerStats(player_id, game_id, team_id)$CA[2], 10)
})


tests/testthat/test.db.R
context("Db tests")
library(nhlscrape)
SetDbPath(example = TRUE)

test_that("We can connect to the database", {
  expect_equal(QueryDb("SELECT * FROM events WHERE game_id=2019020001")$result_event[1], "Game Scheduled")
})


R/db_retrieval.R
#' Send query to database
#'
#' Send a SQL query to the database. Retruns the SQL result as a data.frame. Useful for
#' seeing the structure of the database for construction of queries
#'
#' @param query A string containing the SQL query
#'
#' @examples
#' SetDbPath(example = TRUE)
#' AddGameEvents(2019020001)
#' QueryDb("SELECT * FROM events")
#' QueryDb("SELECT result_description FROM events WHERE game_id=2019020001 AND player_id=8475166")
#'
#' @return List, contains the SQL query result.
#'
#' @export
QueryDb <- function(query) {
  if (nhlscrape.globals$user_set_db == FALSE){
    stop("No user defined database found. Please use SetDbPath().")
  }
  conn <- DBI::dbConnect(RSQLite::SQLite(), nhlscrape.globals$db_file)
  record <- DBI::dbGetQuery(conn, query)
  DBI::dbDisconnect(conn)

  return(record)
}

#' Retrieve database location on system
#'
#' Returns the current path to the local database file
#'
#' @examples
#' GetDbPath()
#'
#' @return Character, system path to the database.
#'
#' @export
GetDbPath <- function() {
  return(nhlscrape.globals$db_file)
}

#' Set database location on system
#'
#' Sets the path to the database file. This function must be run to allow operation
#' of other functions which require saving to a database
#'
#' @param db_path Character, contains the system path to the db file
#' @param example boolean, set to FALSE by default, if TRUE it will
#' copy the example db file and set the path pointing to it
#'
#' @examples
#' SetDbPath(example = TRUE)
#'
#' @return Character, contains the path to the database.
#'
#' @export
SetDbPath <- function(db_path="nhl.sqlite", example=FALSE) {
  if (example) {
    test_folder = tempdir()
    file.copy(system.file("extdata", "nhl.sqlite", package = "nhlscrape"), test_folder)
    db_path = file.path(test_folder, "nhl.sqlite")
  }
  nhlscrape.globals$user_set_db <-TRUE
  nhlscrape.globals$db_file <- db_path
}

#' @keywords internal
#' Check if the events table exists, returns boolean
EventsExists <- function() {
  conn <- DBI::dbConnect(RSQLite::SQLite(), nhlscrape.globals$db_file)
  result <- DBI::dbListTables(conn)
  DBI::dbDisconnect(conn)

  return("events" %in% result)
}

#' Search for team id from team name
#'
#' Retrieve the team ID using the abbreviation, or full name of the team
#'
#' @param team_name Character, contains either abbreviation or full name
#'
#' @return Int, team ID number.
#'
#' @examples
#' SetDbPath(example = TRUE)
#' AddAllTeamsDb()
#' GetTeamId("TOR")
#' GetTeamId("tor")
#' GetTeamId("Toronto Maple Leafs")
#'
#' @export
GetTeamId <- function(team_name) {
  # Expect name to be either full name or abbreviation
  team_id <- rbind(QueryDb(paste("SELECT * FROM teams WHERE name='", team_name, "'", sep="")),
                   QueryDb(paste("SELECT * FROM teams WHERE UPPER(abbreviation)='", toupper(team_name), "'", sep=""))
  )
  if (nrow(team_id) == 0) {
    stop("Could not find team with name: ", team_name)
  }
  return(team_id$pk)
}

#' @keywords internal
GetTeamRoster <- function(team_id, year) {
  request <- paste("teams/", team_id, "/roster", sep="")
  r <- GetApiJson(request)

  roster <- r$roster
  return(c(roster$person,
           data.frame("jerseyNumber" = roster$jerseyNumber, stringsAsFactors = FALSE),
           roster$position))
}

#' @keywords internal
GetPlayerStatsYears <- function(player_id, year_range) {
  request <- paste("people/",
                   player_id,
                   "/stats?stats=statsSingleSeason&season=",
                   year_range,
                   sep="")
  r <- GetApiJson(request)
  return(r$stats$splits[[1]]$stat)
}

#' @keywords internal
GetGameIdNext <- function(team_id) {
  request <- paste("people/", team_id, "?expand=team.schedule.next", sep="")
  r <- GetApiJson(request)
  return(r$teams$nextGameSchedule$dates[[1]]$games[[1]]$gamePk)
}

#' @keywords internal
GetGameIdPrevious <- function(team_id) {
  request <- paste("teams/", team_id, "?expand=team.schedule.previous", sep="")
  r <- GetApiJson(request)
  return(r$teams$previousGameSchedule$dates[[1]]$games[[1]]$gamePk)
}

#' Find game ids for a specific team and date range
#'
#' Gets a list of game ids for team_id in a specific date range, inclusive
#'
#' @param team_id Int, team ID number
#' @param start_date Starting date of the games, inclusive. Format: "yyyy-mm-dd"
#' @param end_date Ending date of the games, inclusive. Format: "yyyy-mm-dd"
#'
#' @return List of ints, each element is a game ID in selected range
#'
#' @examples
#' \donttest{
#' GetGameIdRange(10, "2019-09-30", "2019-12-16")
#' }
#' @export
GetGameIdRange <- function(team_id, start_date, end_date) {
  request <- paste("schedule?teamId=", team_id, sep="")
  request <- paste(request, "&startDate=", start_date, sep="")
  request <- paste(request, "&endDate=", end_date, sep="")
  r <- GetApiJson(request)

  game_ids <- c()
  for (game in r$dates$games) {
    game_ids <- c(game_ids, game$gamePk)
  }

  return(game_ids)
}

#' @keywords internal
#' Helper used to get the player id from their jersey number in a specific game.
GetPlayerIdFromNumber <- function(number, player_list) {
  player_id <- NA
  if (is.na(number)) {
    return(player_id)
  }
  for (player in player_list) {
    if (number == player$jerseyNumber) {
      player_id <- player$person$id
    }
  }
  return(player_id)
}

#' Search player id using name
#'
#' Gets a player id from their name. Will only work for players that were active in a game that
#' has already been added to the database
#'
#' @param player_name character, players full name
#'
#' @return int, player id number
#'
#' @examples
#' SetDbPath(example = TRUE)
#' AddGameEvents(2019020001)
#' GetPlayerId("John Tavares")
#'
#' @export
GetPlayerId <- function(player_name) {
  query <- paste("SELECT id FROM players WHERE UPPER(fullName)='", toupper(player_name), "'",
                 sep="")
  return(QueryDb(query)$id)
}

#' @keywords internal
#' Helper function for checking if a play is even strength, checks the goalies
#' and the total player count, returns boolean
IsEven <- function(row) {
  if(is.na(row["players_on_ice"])) {
    return(FALSE)
  }
  # Goalie in net check
  home_goalie <- as.logical(row["home_goalie"])
  visitor_goalie <- as.logical(row["visitor_goalie"])

  plrs <- strsplit(row["players_on_ice"], ",")[[1]]
  plrs <- setdiff(plrs, "NA")
  is_even <- length(plrs) == 12 && home_goalie && visitor_goalie
  return(is_even)
}

#' @keywords internal
#' Helper function to cut out rows based on conditional function, ex. IsEven
CutRows <- function(rows, fun) {
  return(rows[apply(rows, 1, fun),])
}

#' Get the heatmap coordinates for certain events in the database
#'
#' This function will take in a team_id and a list of game_ids, and a list of
#' events to looks for, and resturn a list of x, y coordinates, transformed to
#' be attacking zone on the left of the rink, and defending zone on the right of the rink.
#' Some anomalies due to how the NHL tracks positions
#'
#' @param team_id Int, id of the team to transfrom for
#' @param gids Int, list of game ids to check for events
#' @param events_list Character, string of events to select from the database
#' Note that events must be quoted in single quotes inside the string
#'
#' @return List, containing all x, y pairs for events and additional metadata
#'
#' @examples
#' SetDbPath(example = TRUE)
#' GetHeatmapCoords(10, 2019020001, "'Shot', 'Goal'")
#'
#' @export
GetHeatmapCoords <- function(team_id, gids, events_list) {
  all_rows <- 0
  for (game_id in gids) {
    #query <- paste("SELECT home_team, away_team FROM game_info WHERE game_id=", game_id, sep="")
    #row <- QueryDb(query)

    #if (GetTeamId(row$home_team) == team_id) {
    #  team_home <- TRUE
    #} else {
    #  team_home <- FALSE
    #}

    query <- paste("SELECT coordinates_x FROM events WHERE game_id=", game_id, " AND team_id=", team_id,
                   " AND result_event='Shot' AND about_period=1", sep = "")
    row <- QueryDb(query)
    avg = mean(row$coordinates_x)

    if (avg < 0) {
      flip <- TRUE
    } else {
      flip <- FALSE
    }

    query <- paste("SELECT DISTINCT coordinates_x,
    coordinates_y,
    result_description,
    about_period,
    about_periodTime,
    about_dateTime,
    game_id,
                   result_event FROM events WHERE game_id=", game_id,
                   " AND team_id=", team_id,
                   " AND result_event IN (", events_list, ")",
                   sep="")
    rows <- QueryDb(query)
    for (i in 1:nrow(rows)) {
      if (flip && rows[i,]$about_period %% 2 == 0) {
        # Flip second period
        rows[i,]$coordinates_x <- -1 * rows[i,]$coordinates_x
      } else if (!flip && rows[i,]$about_period %% 2 == 1) {
        # Flip first and last period
        rows[i,]$coordinates_x <- -1 * rows[i,]$coordinates_x
      }
    }
    if (typeof(all_rows) == "double") {
      all_rows <- rows
    } else {
      all_rows <- rbind(all_rows, rows)
    }
  }
  return(all_rows)
}

#' WIP - Get advanced statistics for player_id on team_id in a list of games.
#'
#'
#' This function will search all games in game_ids and return a list of stats for
#' the player id selected. The current stats returned are:
#' - Shots
#' - Goals
#' - Corsi
#' - Fenwick
#'
#' @param player_id Int, player ID number
#' @param game_ids List, game ids to check. Must already be in the database
#' @param team_id Int, the ID of the team the player plays for
#'
#' @return List, contains a row of stats for even strength and for all situations
#'
#' @examples
#' SetDbPath(example = TRUE)
#' AddGameEvents(2019020001)
#' GetPlayerStats(8475166, 2019020001, 10)
#'
#' @export
GetPlayerStats <- function(player_id, game_ids, team_id) {

  num_stats <- 3
  num_situations <- 2
  # Initialize stats df
  stats <- data.frame(matrix(ncol = num_stats, nrow = num_situations))
  names(stats) <- c("CF", "CA", "S")
  rownames(stats) <- c("All_situations", "Even_strength")


  gids_str <- paste(game_ids, collapse=",")

  # Corsi ----------------------------------------------------------------
  # CF in all situations
  query <- paste("SELECT * FROM events WHERE game_id IN (", gids_str, ")",
                 " AND (playerType='Shooter' OR playerType='Scorer')",
                 " AND players_on_ice LIKE '%", player_id, "%'",
                 " AND player_team_id='", team_id, "'",
                 sep="")
  CF_all <- QueryDb(query)

  # CA in all situations
  query <- paste("SELECT * FROM events WHERE player_id!=", player_id,
                 " AND game_id IN (", gids_str, ")",
                 " AND (playerType='Shooter' OR playerType='Scorer')",
                 " AND players_on_ice LIKE '%", player_id, "%'",
                 " AND player_team_id!='", team_id, "'",
                 sep="")
  CA_all <- QueryDb(query)

  # CF at even strength
  CF_even <- CutRows(CF_all, IsEven)

  # CA at even strength
  CA_even <- CutRows(CA_all, IsEven)

  stats["All_situations", c("CF", "CA")] <- c(nrow(CF_all), nrow(CA_all))
  stats["Even_strength", c("CF", "CA")] <- c(nrow(CF_even), nrow(CA_even))

  # Shots ----------------------------------------------------------------
  # Shots All Situations
  query <- paste("SELECT * FROM events WHERE game_id IN (", gids_str, ")",
                 " AND player_id=", player_id,
                 " AND (playerType='Shooter' OR playerType='Scorer')",
                 " AND (result_eventTypeId='SHOT' OR result_eventTypeId='GOAL')",
                 " AND about_periodType!='SHOOTOUT'",
                 sep="")
  S_all <- QueryDb(query)

  # Shots at even strength
  S_even <- CutRows(S_all, IsEven)


  stats["All_situations", "S"] <- nrow(S_all)
  stats["Even_strength", "S"] <- nrow(S_even)

  # Fenwic ----------------------------------------------------------------
  # FF in all situations
  query <- paste("SELECT * FROM events WHERE game_id IN (", gids_str, ")",
                 " AND result_eventTypeId!='BLOCKED_SHOT'",
                 " AND (playerType='Shooter' OR playerType='Scorer')",
                 " AND players_on_ice LIKE '%", player_id, "%'",
                 " AND player_team_id='", team_id, "'",
                 sep="")
  FF_all <- QueryDb(query)

  #FF at even strength NOT MATCHING UP!?!?
  FF_even <- CutRows(FF_all, IsEven)

  #FA in all situations
  query <- paste("SELECT * FROM events WHERE player_id!=", player_id,
                 " AND game_id IN (", gids_str, ")",
                 " AND result_eventTypeId!='BLOCKED_SHOT'",
                 " AND (playerType='Shooter' OR playerType='Scorer')",
                 " AND players_on_ice LIKE '%", player_id, "%'",
                 " AND player_team_id!='", team_id, "'",
                 sep="")
  FA_all <- QueryDb(query)

  #FA at even strength
  FA_even <- CutRows(FA_all, IsEven)

  # Goals ----------------------------------------------------------------
  # GF in all situations
  query <- paste("SELECT * FROM events  WHERE game_id IN (", gids_str, ")",
                 " AND playerType='Scorer'",
                 " AND players_on_ice LIKE '%", player_id, "%'",
                 " AND player_team_id='", team_id, "'",
                 sep="")
  GF_all <- QueryDb(query)

  #GF at even strength
  GF_even <- CutRows(GF_all, IsEven)

  # GA in all situations
  query <- paste("SELECT * FROM events  WHERE game_id IN (", gids_str, ")",
                 " AND playerType='Scorer'",
                 " AND players_on_ice LIKE '%", player_id, "%'",
                 " AND player_team_id!='", team_id, "'",
                 sep="")
  GA_all <- QueryDb(query)

  #GA at even strength
  GA_even <- CutRows(GA_all, IsEven)


  return(stats)
}


R/nhl_scrape.R
#================================================================
# Setup
#================================================================
api_url <- "https://statsapi.web.nhl.com/api/v1/"

# Check if a db exists, if not then create an empty db
db_file <-"nhl.sqlite"
db_location <- "data_raw/"
db_path <- paste(db_location, db_file, sep="")
if (!file.exists(db_path)) {
  dir.create(db_location, showWarnings = FALSE)
  file.create(db_path)
}
EVENTS_EMPTY = TRUE

#================================================================
# Functions
#================================================================

# Database manipulation block ----------------------------------------------------------------

#' @keywords internal
#' Check if the record is in the db already
ExistsInDb <- function(table, pk) {
  query <- paste("SELECT * FROM ", table, " WHERE pk='", pk, "'", sep="")
  record <- QueryDb(query)

  if (nrow(record) == 0) {
    return(FALSE)
  }
  return(TRUE)
}

#' @keywords internal
#' Add the dataframe to the database under 'table'
AddDb <- function(table, df) {
  # Add all rows to db, checking if the exist already
  conn <- DBI::dbConnect(RSQLite::SQLite(), db_path)

  # Create the table if it doesn't exist yet (keep 0 rows)
  if (!DBI::dbExistsTable(conn, table)) {
    DBI::dbWriteTable(conn, table, df[FALSE,])
  }

  # Append the row if it doesn't exist yet
  for (i in 1:nrow(df)) {
    row <- df[i,]
    pk <- row$pk
    if (!ExistsInDb(table, pk)) {
      DBI::dbWriteTable(conn, table, row, append=TRUE, row.names=FALSE)
    }
  }
  DBI::dbDisconnect(conn)

  message(paste("'", table, "' rows added successfully", sep=""))
}

#' Send a SQL query to the database.
#'
#' @param query A string containing the SQL query.
#' @examples
#' QueryDb("SELECT * FROM events")
#' QueryDb("SELECT player_id, player_fullName FROM events WHERE game_id=2019020001")
QueryDb <- function(query) {
  conn <- DBI::dbConnect(RSQLite::SQLite(), db_path)
  record <- DBI::dbGetQuery(conn, query)
  DBI::dbDisconnect(conn)

  return(record)
}

#' Add all teams to the database.
#'
#' @examples
#' AddAllTeamsDb()
AddAllTeamsDb <- function() {
  df <- jsonlite::flatten(GetApiJson("teams")$teams)

  # Make names db friendly
  cols <- gsub("\\.", "_", names(df))
  cols <- gsub("id", "pk", cols)
  names(df) <- cols

  # Remove duplicate id
  df <- within(df, rm("franchise_franchiseId"))

  # Add to database
  AddDb("teams", df)
}

#' Add a teams roster to the 'roster' table for 'season'.
#'
#' @param team_id Identity number for the team. Use GetTeamId to find.
#' @param season A year range you want to add.
#' @examples
#' AddTeamRoster(10, 20192020)
#' AddTeamRoster(3, 20142015)
AddTeamRoster <- function(team_id, season) {
  request <- paste("teams/", team_id, "?expand=team.roster&season=", season, sep="")
  df <- jsonlite::flatten(GetApiJson(request)$teams$roster$roster[[1]])

  # Make names db friendly
  cols <- gsub("\\.", "_", names(df))
  names(df) <- cols

  # Make a new col for the pk
  df$pk <- paste(df$person_id, season, sep="_")
  df$team_id <- as.integer(team_id)

  # Add to database
  AddDb("rosters", df)
}

#' Add all events from a 'game_id' to the 'events' table
#'
#' @param game_id Identifying number for the game. Use GetGameIds to find.
#' @examples
#' AddGameEvents(2019020357)
#' AddGameEvents(2019020405)
AddGameEvents <- function(game_ids) {
  for (game_id in game_ids) {
    if (EVENTS_EMPTY == FALSE && nrow(QueryDb(paste("SELECT * FROM events WHERE game_id=", game_id))) > 0) {
      message(paste("game with game_id:'", game_id,"'already in database", sep=""))
      next
    }
    EVENTS_EMPTY = FALSE

    request <- paste("game/", game_id, "/feed/live", sep="")
    df <- GetApiJson(request)

    # Need this for constant roster data
    request <- paste("game/", game_id, "/boxscore", sep="")
    boxscore <- GetApiJson(request)
    home_roster <- boxscore$teams$home$players
    away_roster <- boxscore$teams$away$players


    players <- df$liveData$plays$allPlays$players
    plays <- df$liveData$plays$allPlays
    plays <- within(plays, rm("players"))

    home_abbr <- df$gameData$teams$home$triCode
    visitor_abbr <- df$gameData$teams$away$triCode

    # We need to get the html report of the game so we an tell who was on the ice on any given play
    # All we need to do is match the times to any time in the report and insert the players
    html_report <- GetGameLiveFeedHtml(game_id)

    df_final <- data.frame()
    for (i in 1:length(players)) {
      row <- players[[i]]

      if (!is.null(row)) {
        for (j in 1:nrow(row)) {
          player <- row[j,]
          event <- plays[i,]

          # Now combine all info with the game id for the pk
          df <- jsonlite::flatten(cbind(player, event))
          cols <- gsub("\\.", "_", names(df))
          names(df) <- cols

          # Check if we have the seasonTotal field missing
          if (!("seasonTotal" %in% cols)) {
            df <- cbind(df, "seasonTotal"=NA)
          }

          # Finally make the pk field
          df$pk <- paste(game_id, df$about_eventIdx, df$player_id, sep="_")

          # Append player current team id for ident
          player_list <- c(home_roster, away_roster)
          player_id_num <- paste("ID", row$player$id[[j]], sep="")
          player_team_id <- player_list[[player_id_num]]$person$currentTeam$id
          df <- cbind(df, "player_team_id"=player_team_id)

          # Get time stamp for player on ice identification
          period <- df$about_period
          time_period <- df$about_periodTime
          # Remove first zero to match html format
          if (substring(time_period, 1, 1) == "0"){
            time_period <- substring(time_period, 2)
          }
          tmp_event <- subset(html_report, period_html==period & time_elapsed==time_period)
          # Select first row
          tmp_row <- tmp_event[1,]

          # Add each player id to the current row
          df <- cbind(df, "players_on_ice"=paste(
                      GetPlayerIdFromNumber(tmp_row$visitor_p1, away_roster),
                      GetPlayerIdFromNumber(tmp_row$visitor_p2, away_roster),
                      GetPlayerIdFromNumber(tmp_row$visitor_p3, away_roster),
                      GetPlayerIdFromNumber(tmp_row$visitor_p4, away_roster),
                      GetPlayerIdFromNumber(tmp_row$visitor_p5, away_roster),
                      GetPlayerIdFromNumber(tmp_row$visitor_p6, away_roster),
                      GetPlayerIdFromNumber(tmp_row$home_p1, home_roster),
                      GetPlayerIdFromNumber(tmp_row$home_p2, home_roster),
                      GetPlayerIdFromNumber(tmp_row$home_p3, home_roster),
                      GetPlayerIdFromNumber(tmp_row$home_p4, home_roster),
                      GetPlayerIdFromNumber(tmp_row$home_p5, home_roster),
                      GetPlayerIdFromNumber(tmp_row$home_p6, home_roster), sep=",") )

          # Add goalie booleans from home and visitor
          df <- cbind(df, "home_goalie"=tmp_row$home_goalie)
          df <- cbind(df, "visitor_goalie"=tmp_row$visitor_goalie)

          df_final <- rbind(df_final, df)

        }
      } else {
        event <- plays[i,]

        df <- jsonlite::flatten(event)
        cols <- gsub("\\.", "_", names(df))
        names(df) <- cols
        df$pk <- paste(game_id, df$about_eventIdx, sep="_")

        # Add NA entries so column numbers match
        df <- cbind(df,
                    "playerType"=NA,
                    "player_id"=NA,
                    "player_team_id"=NA,
                    "player_fullName"=NA,
                    "player_link"=NA,
                    "seasonTotal"=NA,
                    "players_on_ice"=NA,
                    "home_goalie"=FALSE,
                    "visitor_goalie"=FALSE)
        df_final <- rbind(df_final, df)
      }

    }
    # Finally, add game_id column
    df_final <- cbind(df_final, game_id=rep(game_id, nrow(df_final)))

    AddDb("events", df_final)
    message(paste("for game_id:'", game_id, "'", sep=""))
  }
}


# Database retrieval block ----------------------------------------------------------------

#' Retrieve the team ID using the abbreviation, or full name of the team.
#'
#' @param team_name String containing either abbreviation or full name.
#'
#' @return Int, team ID number.
#'
#' @examples
#' GetTeamId("TOR")
#' GetTeamId("Toronto Maple Leafs")
GetTeamId <- function(team_name) {
  # Expect name to be either full name or abbreviation
  team_id <- rbind(QueryDb(paste("SELECT * FROM teams WHERE name='", team_name, "'", sep="")),
                   QueryDb(paste("SELECT * FROM teams WHERE abbreviation='", team_name, "'", sep="")),
                   QueryDb(paste("SELECT * FROM teams WHERE name_upper='", team_name, "'", sep=""))
  )
  if (nrow(team_id) == 0) {
    stop("Could not find team with name: ", team_name)
  }
  return(team_id$id)
}

GetTeamRoster <- function(team_id, year) {
  request <- paste("teams/", team_id, "/roster", sep="")
  r <- GetApiJson(request)

  roster <- r$roster
  return(c(roster$person,
           data.frame("jerseyNumber" = roster$jerseyNumber, stringsAsFactors = FALSE),
           roster$position))
}

GetPlayerStatsYears <- function(player_id, year_range) {
  request <- paste("people/",
                   player_id,
                   "/stats?stats=statsSingleSeason&season=",
                   year_range,
                   sep="")
  r <- GetApiJson(request)
  return(r$stats$splits[[1]]$stat)
}

GetGameIdNext <- function(team_id) {
  request <- paste("people/", team_id, "?expand=team.schedule.next", sep="")
  r <- GetApiJson(request)
  return(r$teams$nextGameSchedule$dates[[1]]$games[[1]]$gamePk)
}

GetGameIdPrevious <- function(team_id) {
  request <- paste("teams/", team_id, "?expand=team.schedule.previous", sep="")
  r <- GetApiJson(request)
  return(r$teams$previousGameSchedule$dates[[1]]$games[[1]]$gamePk)
}

#' Gets a list of game ids for team_id in a specific date range.
#'
#' @param team_id Team ID number.
#' @param start_date Starting date of the games, inclusive. Format: "yyyy-mm-dd".
#' @param end_date Ending date of the games, inclusive. Format: "yyyy-mm-dd".
#'
#' @return List of ints, each element is a game ID in selected range.
#'
#' @examples
#' GetGameIdRange(10, "2019-09-30", "2019-12-16")
GetGameIdRange <- function(team_id, start_date, end_date) {
  request <- paste("schedule?teamId=", team_id, sep="")
  request <- paste(request, "&startDate=", start_date, sep="")
  request <- paste(request, "&endDate=", end_date, sep="")
  r <- GetApiJson(request)

  game_ids <- c()
  for (game in r$dates$games) {
    game_ids <- c(game_ids, game$gamePk)
  }

  return(game_ids)
}

#' @keywords internal
#' Helper used to get the player id from their jersey number in a specific game.
GetPlayerIdFromNumber <- function(number, player_list) {
  player_id <- NA
  if (is.na(number)) {
    return(player_id)
  }
  for (player in player_list) {
    if (number == player$jerseyNumber) {
      player_id <- player$person$id
    }
  }
  return(player_id)
}

#' @keywords internal
#' Helper function for checking if a play is even strength, checks the goalies
#' and the total player count, returns boolean
IsEven <- function(row) {
  # Goalie in net check
  home_goalie <- as.logical(row$home_goalie)
  visitor_goalie <- as.logical(row$visitor_goalie)

  plrs <- strsplit(row$players_on_ice, ",")[[1]]
  plrs <- setdiff(plrs, "NA")
  is_even <- length(plrs) == 12 && home_goalie && visitor_goalie
  return(is_even)
}

#' WIP - Get advanced statistics for player_id on team_id in a list of games.
#'
#' @param player_id Player ID number.
#' @param game_ids List of game ids to check. Must already be in the database.
#' @param team_id The ID of the team the player plays for.
#'
#' @return Dataframe containing a row of stats for even strength and for all situations.
#'
#' @examples
#' GetPlayerStats(8475166, 2019020001, 10)
GetPlayerStats <- function(player_id, game_ids, team_id) {

  # Initialize stats df
  stats <- data.frame(matrix(ncol = 3, nrow = 0))
  corsi_for_all <- 0
  corsi_against_all <- 0
  corsi_for_even <- 0
  corsi_against_even <- 0

  for (game_id in game_ids) {

    # Corsi
    #----------------------------------------------------------------
    # CF all situations
    query <- paste("SELECT * FROM events WHERE game_id=", game_id,
                   " AND (playerType='Shooter' OR playerType='Scorer')",
                   " AND players_on_ice LIKE '%", player_id, "%'",
                   " AND player_team_id='", team_id, "'",
                   sep="")
    rows <- QueryDb(query)
    corsi_for_all <- corsi_for_all + nrow(rows)

    # CF in even strength situations
    for (i in 1:nrow(rows)) {
      if(i == 0) {
        next
      }
      row <- rows[i,]

      if (IsEven(row)) {
        corsi_for_even <- corsi_for_even + 1
      }
    }

    # CA in all situations
    query <- paste("SELECT * FROM events WHERE player_id!=", player_id,
                   " AND game_id=", game_id,
                   " AND (playerType='Shooter' OR playerType='Scorer')",
                   " AND players_on_ice LIKE '%", player_id, "%'",
                   " AND player_team_id!='", team_id, "'",
                   sep="")
    rows <- QueryDb(query)
    corsi_against_all <- corsi_against_all + nrow(rows)

    # CA at even strength
    for (i in 1:nrow(rows)) {
      if(i == 0) {
        next
      }
      row <- rows[i,]

      if (IsEven(row)) {
        corsi_against_even <- corsi_against_even + 1
      }
    }


  }


  # Finalize stats df
  # Corsi
  corsi_all <- corsi_for_all - corsi_against_all
  corsi_even <- corsi_for_even - corsi_against_even


  stats <- rbind(stats, c(corsi_for_all, corsi_against_all, corsi_all), c(corsi_for_even, corsi_against_even, corsi_even))
  names(stats) <- c("CF", "CA", "C")
  rownames(stats) <- c("All_situations", "Even_strength")
  return(stats)
}

# Scraping block (api and html) ----------------------------------------------------------------

#' @keywords internal
#' Calls the api with call, returns the html
GetApiJson <- function(call) {
  request <- paste(api_url, call, sep="")
  r <- httr::GET(request)

  # Logging
  log <- paste("[", Sys.time(), "] ", request, sep="")
  write(log, file="requests.log", append=TRUE)

  # Make sure we have the correct data from the GET request
  httr::stop_for_status(r)

  # Get the text and parse it from JSON to a table
  txt <- httr::content(r, "text")
  json <- jsonlite::fromJSON(txt)
  return(json)
}

#' @keywords internal
#' Scrapes and returns the html report for game_id
GetGameLiveFeedHtml <- function(game_id) {
  # Get years in correct format
  year <- substring(game_id, 1, 4)
  next_year <- as.character(as.integer(year) + 1)
  years <- paste(year, next_year, sep="")

  id <- paste("PL", substring(game_id, 5), sep="")

  url <- paste("http://www.nhl.com/scores/htmlreports/", years, "/", id, ".HTM", sep="")

  content <- xml2::read_html(url)

  plays_data <- rvest::html_nodes(content, xpath='//*[contains(concat( " ", @class, " " ), concat( " ", "bborder", " " ))]')
  visitor_abbr <- substring(rvest::html_text(plays_data[7]), 1, 3)
  home_abbr <- substring(rvest::html_text(plays_data[8]), 1, 3)

  col <- 1
  row <- c()
  options(stringsAsFactors=FALSE)
  data <- data.frame()
  for (i in 1:length(plays_data)) {

    # If col 4 (time), we need to take only the time elapsed
    if (col == 4) {
      str <- as.character(plays_data[i])
      str <- gsub("<br>", "\n", str)
      val <- rvest::html_text(xml2::read_html(str))
      val <- strsplit(val, "\n")[[1]][1]
    } else if (col == 7) {
      # Parse players into lists
      str <- rvest::html_text(plays_data[i])
      plrs <- strsplit(str, "\r\n\r\n")[[1]]

      away_p1 <- substring(plrs[2], 3)
      away_p2 <- substring(plrs[5], 3)
      away_p3 <- substring(plrs[8], 3)
      away_p4 <- substring(plrs[11], 3)
      away_p5 <- substring(plrs[14], 3)
      away_p6 <- substring(plrs[17], 3)

      # Add a check for the goalies on the ice for 5v5 stat calculations
      goalie <- FALSE
      check <- plrs[3] == "G" ||
        plrs[6] == "G" ||
        plrs[9] == "G" ||
        plrs[12] == "G" ||
        plrs[15] == "G" ||
        plrs[18] == "G"
      if (!is.na(check) && check) {
        goalie <- TRUE
      }

      row <- c(row, goalie, away_p1, away_p2, away_p3, away_p4, away_p5)
      val <- away_p6

    } else if (col == 8) {
      # Parse players into lists
      str <- rvest::html_text(plays_data[i])
      plrs <- strsplit(str, "\r\n\r\n")[[1]]

      home_p1 <- substring(plrs[2], 3)
      home_p2 <- substring(plrs[5], 3)
      home_p3 <- substring(plrs[8], 3)
      home_p4 <- substring(plrs[11], 3)
      home_p5 <- substring(plrs[14], 3)
      home_p6 <- substring(plrs[17], 3)

      # Add a check for the goalies on the ice for 5v5 stat calculations
      goalie <- FALSE
      check <- plrs[3] == "G" ||
        plrs[6] == "G" ||
        plrs[9] == "G" ||
        plrs[12] == "G" ||
        plrs[15] == "G" ||
        plrs[18] == "G"
      if (!is.na(check) && check) {
        goalie <- TRUE
      }

      row <- c(row, goalie, home_p1, home_p2, home_p3, home_p4, home_p5)
      val <- home_p6

    } else {
      val <- rvest::html_text(plays_data[i])
    }

    row <- c(row, val)

    # Reset the column number every 8 nodes, and append row
    col <- col + 1
    if (col == 9) {
      data <- rbind(data, row)
      row <- c()
      col <- 1
    }
  }
  names(data) <- c("ev_html", "period_html", "strength_htm", "time_elapsed", "event_html", "description_html",
                   "visitor_goalie", "visitor_p1", "visitor_p2", "visitor_p3", "visitor_p4", "visitor_p5", "visitor_p6",
                   "home_goalie", "home_p1", "home_p2", "home_p3", "home_p4", "home_p5", "home_p6")
  data <- data[-1,]
  return(data)
}
